### PR TITLE
[SQL] Correct the default size of TimestampType and expose NumericType

### PR DESCRIPTION
**Pull request by yhuai**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 1]
```diff
@@ -402,7 +402,7 @@ case object DateType extends NativeType {
 }
 
 
-protected[sql] abstract class NumericType extends NativeType with PrimitiveType {
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Why do you remove the `protected[sql]`?

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Seems there are users who need it (https://issues.apache.org/jira/browse/SPARK-5288).

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: OK, I see. thanks for explanation

**Comment by marmbrus** [comment]
Thanks!  Merged to master and 1.2

**Pull request closed without merging**